### PR TITLE
feat(Brush API): `Observable`, `clipPath`

### LINKED ISSUE #5431: clipPath for canvas ignores free drawing
## Version
2.4.5

## Test Case
http://jsfiddle.net/synamqvu/

## Steps to reproduce
Try to draw anything, free drawing mode enabled.

## Expected Behavior
A circle is used as clipPath for canvas, it should clip contextTop as well.
## Actual Behavior
It is clipping only after mouse up. clipPath is not applying to contextTop while using free drawing.

### LINKED ISSUE #6955: Adding Freehand with ClipPath to Object Group
Hello,

I am working with Fabricjs and i'd like to draw inside of a circle and have the path constrained to the circle. This includes the selection area of the group.

UPDATE:

Seem to have solved the first point

http://jsfiddle.net/42snph1d/14/

**However, the selection are is too big. Anyway to reduce it so it is the side of the circle and not the path?**

**Is it possible to truncate the PATH to only include points inside the circle?**

--------------------------------------------------------------------------------------------------------------

http://jsfiddle.net/42snph1d/12/

I am seeing weird artefacts when moving the object around. 

1) The freehand drawing can be seen in the canvas outside of the circle when moving
2) The group selection is large, includes the entire freehand drawing. I'd like the selection area to only contain the circle.

Any obvious fixes here?

circle

Any advice would be appreciated.

### PR DESCRIPTION
**Pull request by ShaMan123**
## Motivation

closes #7175 
closes #7134 
closes #6955 
closes #5431

## Description

This PR is a port of #7175.
It sets sail to the new Brush API, which IMO should not be called brush anymore.
As part of the new API, brush now extends `Observable` and is subscribed to all canvas events by default.
I believe this change is so good that it will unlock lots of options including moving interaction code (e.g. group selector) to a dedicated brush.
I also think `freeDrawingBrush` should be renamed and become an array of interactions (this will enable supporting multi touch interactions).
To allow all this I have exposed a long desired feature, `FabricEvent#preventDefault`. Brushes listen to the before event (of down, move, up) and prevent default in case they operate and then canvas knows not to run the default action. This is great since there are interactions that should not prevent default behavior such as visualizing pointers etc.
In order to make brush flexible in terms of how it interacts with canvas events I have exposed the `subscribe` method that wires the brush logic to the canvas events so subclassing this method is very powerful.

## Changes

**This is a big PR but not a huge one. There are a lot of goldens changed.**

API changes:

- Brush subclasses Observable
- Observable now fires a `FabricEvent` instance to support `preventDefault` and `stopPropagation`
- Brush has a `subscribe` method that connects between lifecycle methods to canvas events. This way a brush can start its lifecycle when contextmenu event is fired instead of when mousedown is fired w/o changing logic accept the subscribe method
- Created `SimpleBrush`: a subclass of `BaseBrush` that is wired to interact with a simple `down-move-up` lifecycle
- Exposed a common `render` + `_render` method on `BaseBrush` which conforms to Object#render and is in charge of:
   1. Preparing the context (transformation etc.)
   2. calling `_render` which is implemented by each brush
   3. render clipPath
   4. cleanup
- enable/disable brush
- brush doesn't add result to canvas anymore and fires `interaction:completed` instead

Minor Changes:

- Added a `clipPath` property to `BaseBrush` and exposed clip path methods
- Added logic to apply the clip path to the resulting object created by the brush
- Moved common logic to base brush
- private flag `_isCurrentlyDrawing` is now a method `isCurrentlyDrawing` (doesn't have to be private)
- canvas fires `resize` from `setDimensions`, brush listens to it to call `_setBrushStyles` instead of canvas doing it
- canvas down, move, up default behavior has been extracted to dedicated methods for ease coding
- renamed `_onResize` => `_onWindowResize` (confusing)
- moved `_setDimensionsImpl` to `Canvas` instaed of `SelectableCanvas`
- removed `button` from `TPointerEventInfo` as we discussed
- deprecated `Canvas#isFreeDrawing` in favor of a fine grained brush option `enabled`
- Moved freedraw goldens to a dedicated folder and added tests that cover this PR

BREAKING:
  - brush doesn't add result to canvas anymore
  - `path:created`, `before:path:created` events are deprecated, use `interaction:completed` instead
  - `_render` method is now protected, use `render` instead
  - renamed `PencilBrush#convertPointsToSVGPath` => `getPathFromPoints`
  - `Canvas#freeDrawingCursor` => `Brush#cursor`
  - `Canvas#isFreeDrawing` => `brush.enable/disable`

## Gist

- [ ] first merge #8430 https://github.com/fabricjs/fabric.js/compare/poly-brush2...clip-brush
- [ ] TODO: PencilBrush doesn't respect `limitToCanvasSize`

## In Action (from #7175)

~~https://codesandbox.io/s/upbeat-microservice-mkbzs?file=/src/App.tsx~~
[STALE sandbox](https://codesandbox.io/s/fabric-app-forked-9hcztm?fontsize=14&hidenavigation=1&module=%2Fsrc%2FBrushClipPath.tsx&theme=dark)

[Edit fabric_app (forked)](https://codesandbox.io/s/fabric-app-forked-n57e00?fontsize=14&hidenavigation=1&module=%2Fsrc%2FBrushClipPath.tsx&theme=dark)

Pay no attention to eraser bug. It's fixed by the group PRs.
ezgif com-gif-maker (3)

**With Eraser patched up**
ezgif com-gif-maker (4)

image

**`absolutePositioned`** + viewportTransform
ezgif com-gif-maker (6)

### CODE REVIEW

File: src/brushes/base_brush.class.ts [hunk 1]
```diff
@@ -161,6 +169,88 @@ export abstract class BaseBrush {
...
+  }
+
+  /**
+   * @private
+   * @param {CanvasRenderingContext2D} ctx
+   * @param {FabricObject} clipPath
+   */
+  drawClipPathOnCache(ctx: CanvasRenderingContext2D, clipPath: FabricObject) {
+    // TODO: no proto calls
+    FabricObject.prototype.drawClipPathOnCache.call(this, ctx, clipPath);
```
Comment (ShaMan123) [inline src/brushes/base_brush.class.ts:188]: I don't like the proto calls but for now it is what it is

**Review by ShaMan123 — COMMENTED** [review]
clip path is ready!
I have adapted the Eraser code but it is still in a broken state. I will handle it in a follow up. I need it as well so I will push the new API forward with Eraser in mind.
Next step I believe will be extending Observable.

**Comment by ShaMan123** [comment]
## Brush API
- blocking flag: if false canvas will handle the event regardless of brush
- extends Observable: **all** canvas events (including non pointer events) will be retargeted to brush, brush will listen to relevant events and execute logic accordingly. This way the dev can use on/off to fine tune behavior.
- Simultaneous multiple active brushes: 
   - assign a canvas element/context to for each brush to draw on using #7860. This will enable many things e.g. tracking users pointers while drawing. 
   - This will also make it possible to draw stuff in respect to object stacking and will make bg/overlay/topContext redundant.
   - And allow to have multiple active brushes
- Move active selection to a brush including groupSelector?
- Handle animations in a dedicated brush?
- lifecycle (not very opinionated):
  - start
  - step: render a step rather than the entire state
  - end

@asturur thoughts?

**Comment by ShaMan123** [comment]
I am thinking it might be a subclass of Layer

**Comment by ShaMan123** [comment]
Extending Group is a good idea. Even only for exposing one pattern less and sticking to good old Object lifecycle

**Comment by ShaMan123** [comment]
So my question would be:
Can we make Object an imperative renderer? Can we use the cache canvas as the context the "brush" will draw over?

**Comment by ShaMan123** [comment]
@asturur I have encountered a bug in brush design I would like to discuss in order to find the best solution.
Brush respects vpt in a very specific way. Its position shouldn't be affected by vpt but its strokeWidth should?
However if vpt changes while drawing... not good. Position is changed.
I have a POC working with `absolutePointer` but I am not sure I will manage to respect `strokeWidth`. Generally speaking interactions are absolute and are not subject to vpt changes. Am I right to assume that?
Should we break brush and make it not subject to vpt and perhaps expose a transform matrix option, same as object?
Thoughts?
This is holding me back.

**Review by ShaMan123 — COMMENTED** [review]
updated from master

**Comment by ShaMan123** [comment]
This PR is GREAT!

**Comment by asturur** [comment]
@ShaMan123 i think there are some things that i agree with in this PR.
Do you mind if i start slicing them and importing in smaller units?

**Comment by ShaMan123** [comment]
Go ahead
You should first merge #8380, or at least work with the diff against it if you don't want to work on that now.

**Comment by ShaMan123** [comment]
This is was good refactor and provided a promising direction
However it has become stale
If in need I will probably rewrite what I need so that:

- Brush is an event target
- All data is in the viewport
- It is bound to its own layer #7880

**Pull request closed without merging**